File: backend/api/llm/openai/schemas.go
package openai

// OpenAIMessage represents a message in the OpenAI chat format
type OpenAIMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// OpenAIJSONSchema represents a JSON schema for structured output
type OpenAIJSONSchema struct {
	Name   string                 `json:"name"`
	Strict *bool                  `json:"strict,omitempty"`
	Schema map[string]interface{} `json:"schema"`
}

// OpenAIResponseFormat specifies the response format for OpenAI
type OpenAIResponseFormat struct {
	Type       string            `json:"type"` // "json_object", "text", or "json_schema"
	JSONSchema *OpenAIJSONSchema `json:"json_schema,omitempty"`
}

// OpenAIRequest represents a request to OpenAI's chat completions API
type OpenAIRequest struct {
	Model               string                `json:"model"`
	Messages            []OpenAIMessage       `json:"messages"`
	Temperature         *float64              `json:"temperature,omitempty"`
	MaxTokens           *int                  `json:"max_tokens,omitempty"`
	MaxCompletionTokens *int                  `json:"max_completion_tokens,omitempty"`
	ReasoningEffort     *string               `json:"reasoning_effort,omitempty"` // "minimal", "low", "medium", "high"
	TopP                *float64              `json:"top_p,omitempty"`
	StopSequences       []string              `json:"stop,omitempty"`
	ResponseFormat      *OpenAIResponseFormat `json:"response_format,omitempty"`
}

// OpenAIChoice represents a choice in the OpenAI response
type OpenAIChoice struct {
	Index        int           `json:"index"`
	Message      OpenAIMessage `json:"message"`
	FinishReason string        `json:"finish_reason"`
}

// OpenAIUsage represents token usage in the OpenAI response
type OpenAIUsage struct {
	PromptTokens     int `json:"prompt_tokens"`
	CompletionTokens int `json:"completion_tokens"`
	TotalTokens      int `json:"total_tokens"`
}

// OpenAIResponse represents a response from OpenAI's chat completions API
type OpenAIResponse struct {
	ID      string         `json:"id"`
	Object  string         `json:"object"`
	Created int64          `json:"created"`
	Model   string         `json:"model"`
	Choices []OpenAIChoice `json:"choices"`
	Usage   OpenAIUsage    `json:"usage"`
}

File: backend/app/features/llm_playground/schemas.py
"""
LLM Playground Schemas.

Pydantic models for request/response validation and serialization.
"""

from typing import Any

from pydantic import BaseModel, Field


class Message(BaseModel):
    """Chat message."""

    role: str = Field(..., description="Role: 'system', 'user', or 'assistant'")
    content: str = Field(..., description="Message content")


class GenerateRequest(BaseModel):
    """
    Unified LLM generation request.

    Model names use provider prefixes for LiteLLM routing:
    - openai/gpt-4o, openai/o1, openai/o3-mini
    - anthropic/claude-3-5-sonnet-20241022, anthropic/claude-3-7-sonnet-20250219
    - gemini/gemini-2.5-pro, gemini/gemini-2.5-flash
    """

    model: str = Field(
        ...,
        description="Model with provider prefix (e.g., 'gemini/gemini-2.5-flash', 'openai/gpt-4o')",
        pattern=r"^[a-z_]+/[a-z0-9._-]+$"
    )
    messages: list[Message] = Field(..., description="Chat messages")

    # Common generation parameters
    temperature: float | None = Field(None, ge=0.0, le=2.0, description="Sampling temperature")
    max_tokens: int | None = Field(None, gt=0, description="Maximum tokens to generate")
    top_p: float | None = Field(None, ge=0.0, le=1.0, description="Nucleus sampling parameter")
    stop: list[str] | None = Field(None, max_length=4, description="Stop sequences")

    # Reasoning/thinking (unified - LiteLLM translates to provider-specific)
    reasoning_effort: str | None = Field(
        None,
        description="Reasoning effort: 'minimal', 'low', 'medium', 'high'. Auto-translates to provider thinking.",
        pattern="^(minimal|low|medium|high)$"
    )

    # OpenAI-specific (for reasoning models)
    max_completion_tokens: int | None = Field(
        None, gt=0, description="Max completion tokens (OpenAI reasoning models)"
    )

    # JSON mode / Structured outputs
    json_mode: bool = Field(False, description="Enable JSON output mode")
    json_schema: dict[str, Any] | None = Field(None, description="JSON schema for structured output")


class Usage(BaseModel):
    """Token usage information."""

    prompt_tokens: int
    completion_tokens: int
    total_tokens: int


class Choice(BaseModel):
    """Generation choice."""

    index: int
    message: Message
    finish_reason: str | None = None


class GenerateResponse(BaseModel):
    """
    LLM generation response (OpenAI-compatible format).

    LiteLLM returns responses in OpenAI format regardless of provider.
    """

    id: str
    object: str
    created: int
    model: str
    choices: list[Choice]
    usage: Usage | None = None
    reasoning_content: str | None = Field(
        None, description="Thinking/reasoning content for reasoning models"
    )


# Model discovery schemas


class ModelInfo(BaseModel):
    """Model information."""

    id: str = Field(..., description="Full model name with prefix (e.g., 'openai/gpt-4o')")
    provider: str = Field(..., description="Provider name: 'openai', 'anthropic', 'gemini'")
    display_name: str = Field(..., description="Human-readable model name")
    supports_reasoning: bool = Field(False, description="Supports reasoning/thinking")
    supports_vision: bool = Field(False, description="Supports vision/image inputs")
    max_tokens: int | None = Field(None, description="Maximum context tokens")


class ModelsResponse(BaseModel):
    """Model listing response."""

    models: list[ModelInfo]
